### PR TITLE
rpcdaemon: add picohttpparser module

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkworm/silkrpc/http/request_parser.cpp [hunk 1]
```diff
@@ -24,270 +24,80 @@
...
+    auto res = phr_parse_request(begin, slen, &method_name, &method_len, &path, &path_len, &minor_version, headers, &num_headers, last_len);
+    if (res < 0) {
+        return ResultType::bad;
+    }
+
+    bool expect_request = false;
+    bool content_length_present = false;
+
+    for (size_t ii = 0; ii < num_headers; ii++) {
+        if (memcmp(headers[ii].name, "Content-Length", headers[ii].name_len) == 0) {
```
Comment (canepat) [inline silkworm/silkrpc/http/request_parser.cpp:65]: Either use `std::memcmp` here and below or add `using namespace std;` at the top

File: silkworm/silkrpc/http/request_parser.cpp [hunk 2]
```diff
@@ -24,270 +24,80 @@
...
-            return false;
+    if (req.content_length == 0) {
+        return ResultType::good;
     }
-}
 
-inline bool RequestParser::is_digit(int c) {
-    return c >= '0' && c <= '9';
+    req.content.resize(slen - static_cast<size_t>(res));
+    memcpy(req.content.data(), &begin[res], slen - static_cast<size_t>(res));
```
Comment (canepat) [inline silkworm/silkrpc/http/request_parser.cpp:94]: Either use `std::memcpy` here and below or add `using namespace std;` at the top

File: silkworm/silkrpc/http/request_parser.cpp [hunk 3]
```diff
@@ -24,270 +24,80 @@
...
-RequestParser::RequestParser() : state_(method_start) {
-}
+RequestParser::ResultType RequestParser::parse(Request& req, const char* begin, const char* end) {
+    const char* method_name;
+    size_t method_len;
+    const char* path;
+    size_t path_len;
+    int minor_version;
+    struct phr_header headers[100];
+    size_t num_headers;
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/http/request_parser.cpp:40]: ok

File: silkworm/silkrpc/http/request_parser.cpp [hunk 4]
```diff
@@ -24,270 +24,80 @@
...
+
+    num_headers = sizeof(headers) / sizeof(headers[0]);
+    auto res = phr_parse_request(begin, slen, &method_name, &method_len, &path, &path_len, &minor_version, headers, &num_headers, last_len);
+    if (res < 0) {
+        return ResultType::bad;
+    }
+
+    bool expect_request = false;
+    bool content_length_present = false;
+
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/http/request_parser.cpp:63]: The capacity is already initialized in connection() constructor (and the capacity is enough). 
I have inserted only one header (Authorization header)

File: silkworm/silkrpc/http/request_parser.cpp [hunk 5]
```diff
@@ -24,270 +24,80 @@
...
+            req.content.push_back(begin[ii]);
+        }
+        if (req.content.length() < req.content_length)
+            return ResultType::indeterminate;
+        else
+            return ResultType::good;
+    }
+
+    num_headers = sizeof(headers) / sizeof(headers[0]);
+    auto res = phr_parse_request(begin, slen, &method_name, &method_len, &path, &path_len, &minor_version, headers, &num_headers, last_len);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/http/request_parser.cpp:56]: ok

File: silkworm/silkrpc/http/request_parser.cpp (hunk 1 above)
Comment (lupin012) [reply silkworm/silkrpc/http/request_parser.cpp:65]: ok

File: silkworm/silkrpc/http/request_parser.cpp (hunk 2 above)
Comment (lupin012) [reply silkworm/silkrpc/http/request_parser.cpp:94]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**